rfcs/text/0008_service_status.md

```diff
@@ -0,0 +1,251 @@
+- If any required dependency is `degraded`, the plugin's status is `degraded`.
+- If any required dependency is `unavailable`, the plugin's status is `unavailable`.
+- If any required dependency is `fatal`, the plugin's status is `fatal`.
+- If any optional dependency is `degraded`, the plugin's status is `degraded`.
+- If any optional dependency is `unavailable`, the plugin's status is `degraded`.
+- If any optional dependency is `fatal`, the plugin's status is `degraded`.
+- Otherwise, the plugin's status is `available`.
+
+If a plugin never calls the `StatusSetup#set` API, the plugin's statud defaults to the inherited status. This inherited status is also exposed on `StatusSetup#inherited$` in order to allow plugins to leverage this inheritance logic while also providing custom status detail on top of the defaults.
+
+### HTTP responses
+
+As specified in the [_Levels section_](#levels), a service's HTTP endpoints will respond with `503 Unavailable` responses in some status levels.
+
+In both the `fatal` and `unavailable` levels, all of a service's endpoints will return 503s. However, in the `degraded` level, it is up to service authors to decide which endpoints should return a 503. This may be implemented directly in the route handler logic or by using any of the [utilities provided](#status-utilities).
+
+## Status Utilities
+
+Though many plugins should be able to rely on the default status inheritance and associated behaviors, there are common patterns and overrides that some plugins will need. The status service should provide some utilities for these common patterns out-of-the-box.
+
+```ts
+/**
+ * Extension of the main Status API
+ */
+interface StatusSetup {
+  /**
+   * Helpers for expressing status in HTTP routes.
+   */
+  http: {
+    /**
+     * High-order route handler function for wrapping routes with 503 logic based
+     * on a predicate.
+     * 
+     * @remarks
+     * When a 503 is returned, it also includes detailed information from the service's
+     * current `ServiceStatus` including `meta` information.
+     * 
+     * @example
+     * ```ts
+     * router.get(
+     *   { path: '/my-api' }
+     *   unavailableWhen(
+     *     ServiceStatusLevel.degraded,
+     *     async (context, req, res) => {
+     *       return res.ok({ body: 'done' });
+     *     }
+     *   )
+     * )
+     * ```
+     * 
+     * @param predicate When a level is specified, if the plugin's current status
+     *                  level is >= to the severity of the  specified level, route
+     *                  returns a 503. When a function is specified, if that
+     *                  function returns `true`, a 503 is returned.
+     * @param handler The route handler to execute when a 503 is not returned.
+     */
+    unavailableWhen<P, Q, B>(
+      predicate: ServiceStatusLevel |
+        (core: CoreStatus, plugins: Record<string, ServiceStatus>) => boolean,
+      handler: RouteHandler<P, Q, B>
+    ): RouteHandler<P, Q, B>;
```

Comment:
`plugins: Record<string, ServiceStatus>` does that contains only dependencies' statuses, or also the current plugin's ?

Reply:
We could add an additional `self` parameter? What would you expect as a developer?

Reply:
I think that if a plugin is going to manually `set` it's state, it will need to access this 'manual' state from within this helper to be able to properly use it? We could just add an entry with the plugin name instead of the `self` value though?

Reply:
> I think that if a plugin is going to manually set it's state, it will need to access this 'manual' state from within this helper to be able to properly use it? We could just add an entry with the plugin name instead of the self value though?

Yes I agree that info should be in here somewhere rather than making the plugin fetch it from it's own Observable. Consistency is nice here.

This is a pretty minor gripe, but I don't love it when plugins have to rely on their own plugin ID in multiple places. It makes renaming plugins slightly more work. I also think having the current plugin in the `plugins` record slightly departs from our other APIs where this object only contains dependencies.

I personally think a `self` argument as the first argument would be more clear, especially if we want to use a similar utility in Core services themselves.

Reply:
I have added a `self` argument.